fs/unpacker.go

```diff
@@ -46,56 +47,184 @@ type Archive interface {
 }
 
 type layerArchive struct {
-	verifier         *layerVerifier
+	compressed       *asyncVerifier
+	uncompressed     *asyncVerifier
 	decompressStream socicompression.DecompressStream
+	bufferPool       *bufferPool
 }
 
-type layerVerifier struct {
-	compressed   digest.Verifier
-	uncompressed digest.Verifier
+type asyncVerifier struct {
+	v       digest.Verifier
+	waitCh  chan struct{}
+	started bool
 }
 
-func NewLayerArchive(compressedVerifier, uncompressedVerifier digest.Verifier, decompressStream socicompression.DecompressStream) Archive {
+func newAsyncVerifier(v digest.Verifier) *asyncVerifier {
+	return &asyncVerifier{
+		v:       v,
+		waitCh:  make(chan struct{}, 1),
+		started: false,
+	}
+}
+
+func (av *asyncVerifier) AsyncVerify(reader io.ReadCloser) {
+	if av == nil || av.v == nil {
+		return
+	}
+	sig := make(chan struct{}, 1)
+	go func() {
+		close(sig)
+		io.Copy(av.v, reader)
+		close(av.waitCh)
+		reader.Close()
+	}()
+	<-sig
+	av.started = true
+}
+
+func (av *asyncVerifier) Verified(ctx context.Context) bool {
+	if av == nil || av.v == nil || !av.started {
```

Comment:
If it hasn't started I don't think we should say it has been verified. Either error out or wait on it IMO.

Reply:
updated